### PR TITLE
Fix for linting and type issues

### PR DESCRIPTION
**Pull request by owenversteeg**
First - excellent boilerplate you've made here. A big thank you from myself and the rest of the Svelte community! Great stack choice to keep things simple, too.

TLDR: I found a handful of issues in the linting, typechecking and general setup, fixed them, found some type errors and fixed those too. I also wrote a small novel (below) about the process which may be helpful to understand exactly why I made each change. 

I was taking a look through the repo and found a few things you might want to fix. First thing I noticed was that the linting didn’t seem to react to my typos - it seems it’s misconfigured. If I run `npm run lint` locally as the CI does, everything passes with no errors, but in fact there are a handful of errors. For example, in `src/routes/(admin)/account/(menu)/+page.server.ts`, `'fail' is defined but never used`, there's a handful of things that are never reassigned but aren't `const`, that sort of thing - just minor stuff. Still, it's stuff the linting should catch, so I took a look and messed around. I added eslint-config-prettier and @typescript-eslint/eslint-plugin to the dev dependencies and modified the .eslintrc.cjs:

```
-  extends: ['eslint:recommended', 'plugin:svelte/recommended'],
+  extends: ['eslint:recommended', 'plugin:svelte/recommended', 'plugin:@typescript-eslint/recommended', 'prettier'],
+  parser: '@typescript-eslint/parser',
+  plugins: ['@typescript-eslint'],
```

With that, I get 39 errors from `npm run lint` and indeed they are minor stuff. I went in to look at those, but while I was busy I noticed a few more things. When you set up the repo with `npm create svelte@latest my-app` you chose JSDoc instead of Typescript, but the repo is (at least now) mostly TS and no JSDoc. The difference between choosing the two is minor; a few things in .eslintrc.cjs, the naming (but not the contents) of jsconfig.json/tsconfig.json, the use of that file name in package.json, a few dependencies in package.json (@typescript-eslint/eslint-plugin, @typescript-eslint/parser and tslib but latter two only Sveltekit >v2) and the extension of .ts/.js for vite.config.js/vite.config.ts but not the file contents. This is minor, but I figured it’s probably good to keep this set up in the standard TS way given that it uses TS, so I made those changes. I kept the configuration changes you made in 761898b, which allows you to use both TS and Javascript in the repo and have both be checked. 

At this point I was ready to fix the errors, but I realized an entirely different set of errors were showing in VS Code - and only in VS Code. These errors (showing up in VS Code but not `npm run lint`==`eslint .`) were all type errors. So - type errors only in VS Code, lint and some but not all type errors with `eslint .` Why? Well, for starters, I thought that the Svelte extension would run linting - but in fact it does not, you need the ESLint extension. I installed it and threw in the configuration into the .vscode/settings.json so it lints Svelte files automatically. Now I had both linting+type errors in VS Code but only lint errors and a few type errors in the console. Why only some of the type errors? A common type error was in src/routes/(admin)/account/(menu)/billing/+page.svelte and a number of other pages: "Object literal may only specify known properties, and 'data' does not exist in type …..” (error caused by doing <SettingsModule {data}>) 

So why was I only getting a small portion of the type errors to show with `eslint .`? Turns out I was barking up the wrong tree; although the Typescript ESLint plugins are helpful and do catch some type errors such as an unexpected `any`, they do not run any actual type checking (which is obvious in hindsight…) so you need to use `svelte-check` (which is in `npm run check`) in addition to linting. It took me a while to realize this; I went down entirely the wrong path because I saw the handful of type errors eslint did catch and assumed it was running type checking. Nope, those are just type errors caught by linting, and once I went back and looked at them I realized that none of those type errors needed anything but some relatively simple matching to catch. Anyway, this is starting to become a novel - I shouldn’t write PRs while caffeinated - but I’ve added the use of svelte-check to the CI as well, so this way the CI catches type errors as well as linting issues. With the changes you made in 761898b, this use of `svelte-check` will also catch detectable errors in Javascript as well as Typescript. 

I went ahead and separated all of this into three commits to keep things clear: 1) intentionally “breaking the CI” by fixing the linting, CI, and ESLint+VS Code setup so all errors are shown (and thus also visible on Github CI here as a secondary check!), 2) fix all the linting errors, 3) to change things over into a Typescript repo setup and update the README with a few notes, and 4) to go through and fix all the type errors. A few notes about the fourth commit; I modified the return value of the action - what gets put into the form prop and $page.form - to encapsulate those into a key `accountUpdateResult` to make the code more clear and prevent any collisions with future form additions. If you can think of a better way to do this or a shorter name I am all ears! There is also one TODO here (link to src/routes/(admin)/account/api/+page.server.ts line 60, the TODO Supabase thing) and here src/routes/(admin)/account/(menu)/settings/change_password/+page.svelte. 

Hopefully all this is helpful! And - one more time - thanks for the project, really appreciate what you’ve done :)

### CODE REVIEW

**Comment by scosman** [comment]
Oh, this looks most excellent. Thanks for the contribution!

Will take me a while to CR, but generally I'm on board for all this. Having good linting + CI was a goal, looks like you made it great. Will review over the next while, and send any questions/concerns.

File: src/routes/(admin)/account/+layout.server.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export const load: PageServerLoad = async ({
 
   const { data: profile } = await supabase
     .from("profiles")
-    .select(`full_name, website, company_name`)
+    .select(`full_name, website, company_name, id, avatar_url, updated_at`)
```
Comment (scosman) [inline src/routes/(admin)/account/+layout.server.ts:15]: Should just be `select *` if we're now using the DB type? Will make maintenance easier. P2.

File: src/ambient.d.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+declare global {
+  type FormResponse = {
```
Comment (scosman) [inline src/ambient.d.ts:2]: Name is too general, and this is really only used for AccountUpdateResult.

Let's use a type AccountUpdateResult or FormAccountUpdateResult, and make it optional where used (instead of nested optional accountUpdateResult)?

**Review by scosman — APPROVED** [review]
Firstly: huge thank you again. This is really excellent work and a great addition to the project.

Secondly: amazing PR in terms of explaining what you did, and breaking it into commits. Merging in my work past fork. 🧑‍🍳🤌

Notes
- Re:amr - agreed. This is documented, but didn’t find a clean type definition maintained by supabase. No point in making a local type def if the owners of the data won’t update it when they make changes. 
- Re:FormResponse naming/nesting -> that’s the one change I think we should make. Let me know if you agree, and if you want to make it or if I should. Not critical by any means.

Only frustrating part: I have a ton of great work to cherry pick into my forks 🤣.

**Comment by owenversteeg** [comment]
You are very welcome! Happy to help :) 

Yeah, I've come full circle, from the newbie that documents everything to the expert that makes complex, inscrutable changes and back around once I realized that the ELI5-style explanations and documentation are useful to everyone. I could tell you were a fan of the same style by the level of detail in the README :) I think too many people dismiss it as handholding or dumbing things down, but it really pays dividends when other people don't have to read the tea leaves to understand the reason for something or what you might mean. 

The AMR typedef stuff makes sense, I modified that comment accordingly. I agree with your other comments as well, so made a few minor changes there. 

The linting caught [a type error](https://github.com/owenversteeg/CMSaasStarter/actions/runs/7864874979/job/21457152197) when I merged in your change, so fixed that. 

The FormResponse naming I agree on, and now I found out why I nested that and discovered a bug. Turns out that in `src/routes/(admin)/account/api/+page.server.ts` you return `errorFields` by passing it to `fail()`, [as Sveltekit encourages](https://kit.svelte.dev/docs/form-actions#anatomy-of-an-action). This data is then accessible via the form prop. However, I mistakenly typed the form prop to `FormResponse`... which actually does not fit the data format; `form` is typically `{ errorResponse, errorFields, email, etc etc }`and `FormResponse` is `{ accountUpdateResult: { errorResponse, errorFields, email, etc etc } }`. Oops! Aside from being a stupid bug, this is also a good example of the limits of typechecking with Svelte and static typechecking in general; the form prop is assembled by Sveltekit at runtime and so Typescript knows nothing about what form it should have. I'm not up to speed on the current state of runtime typechecking so not sure if this is even something people commonly do with TS in 2024, but now I'm getting off-topic.

So, the naming and the bug I introduced have been fixed, linting passes, typechecking passes, should be good to go! I do think it'd be good to do a full pass through to make sure I didn't break anything else but should be good to merge otherwise.

File: src/routes/(admin)/account/api/+page.server.ts [hunk 3]
```diff
@@ -71,7 +76,7 @@ export const actions = {
     }
 
     let validationError
-    let errorFields = []
+    const errorFields = []
```
[the remark this answers is not in the document]
Comment (owenversteeg) [reply src/routes/(admin)/account/api/+page.server.ts:80]: Hah, I know right? Feels like there should be a third option for that case

File: src/routes/(admin)/account/api/+page.server.ts [hunk 4]
```diff
@@ -217,14 +227,12 @@ export const actions = {
...
 
-    const session = await getSession()
-
     const { error } = await supabase.from("profiles").upsert({
       id: session?.user.id,
       full_name: fullName,
       company_name: companyName,
       website: website,
-      updated_at: new Date(),
+      updated_at: new Date().toISOString(),
```
[the remark this answers is not in the document]
Comment (owenversteeg) [reply src/routes/(admin)/account/api/+page.server.ts]: Ah, Typescript has a `Date` type, so I just used that, now we have the timezone info back :)

File: src/routes/(admin)/account/+layout.server.ts (hunk 1 above)
Comment (owenversteeg) [reply src/routes/(admin)/account/+layout.server.ts]: Yep, good catch, changed to `select *`. I think the only real reason not do to this is if you store something sensitive on the user profile, which would go against the row-level security model anyway.

File: src/ambient.d.ts (hunk 2 above)
Comment (owenversteeg) [reply src/ambient.d.ts]: Fixed (see comment)

**Comment by owenversteeg** [comment]
Oh, and side note, for whatever reason CI for your repo doesn't seem to run on (external?) pull requests. IIRC that's something you have to configure in your settings page for this repo? Might be good to have, although in this case you have [my CI](https://github.com/owenversteeg/CMSaasStarter/commits/main/) it will probably come in handy at some point.

**Comment by scosman** [comment]
Sanity checked and everything looks good. Merging! 

Thanks again for the contribution.
